### ISSUE TITLE
Some `del`s are uneffective

### ISSUE THREAD

**Issue by kuraga** [harvardnlp/annotated-transformer#126]
In the Notebook, you sometimes do `del` of either a local variable or a function's argument just before `return`.

IMHO it does nothing there. Both of them are garbage collected exactly there, unconditionally.

**Comment by PangLuo** [comment]
I'm not an expert on python garbage collection. This is what I get from ChatGPT:

Is del query Necessary?
In most cases, explicit del statements are not required because Python’s garbage collector (GC) will eventually free the memory for variables that go out of scope. However:
When working with large models or tensors (like in deep learning), waiting for the GC can lead to inefficiencies and OOM errors.
Explicitly calling del gives you precise control over when memory is released, which is particularly important in long-running processes like model training.

**Comment by kuraga** [comment]
Yes but local variables are collected at the `return`, aren't they?
